client/src/components/FilesDialog/FilesDialog.vue

```diff
@@ -1,16 +1,334 @@
+<script setup lang="ts">
+import { library } from "@fortawesome/fontawesome-svg-core";
+import { faCaretLeft } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
+import { BAlert, BButton } from "bootstrap-vue";
+import Vue, { computed, onMounted, ref } from "vue";
+
+import { UrlTracker } from "@/components/DataDialog/utilities";
+import { isSubPath } from "@/components/FilesDialog/utilities";
+import { selectionStates } from "@/components/SelectionDialog/selectionStates";
+import { errorMessageAsString } from "@/utils/simple-error";
+
+import { getGalaxyInstance } from "../../app";
+import { DirectoryRecord, FileRecord, Model, RecordItem } from "./model";
+import { browseRemoteFiles, FilesSourcePlugin, getFileSources, RemoteEntry } from "./services";
+
+import DataDialogSearch from "@/components/SelectionDialog/DataDialogSearch.vue";
+import DataDialogTable from "@/components/SelectionDialog/DataDialogTable.vue";
+import SelectionDialog from "@/components/SelectionDialog/SelectionDialog.vue";
+
+library.add(faCaretLeft);
+
+interface FilesDialogProps {
+    multiple?: boolean;
+    mode?: "file" | "directory";
+    requireWritable?: boolean;
+    callback?: (files: any) => void;
```

Comment:
Since there are default values assigned for each, they are no longer considered optional and have the same type

```suggestion
    multiple: boolean;
    mode: "file" | "directory";
    requireWritable: boolean;
    callback: (files: any) => void;
```

Reply:
Can you elaborate on this? I'm a bit confused, even if all optional properties have a default value (which always should by definition) doesn't mean they stop being optional right? (you don't need to pass them when using the component), making all of them required (and optional because of the default) is confusing to me...

Reply:
@davelopez Since the Composition API and defining default values for the props by `withDefaults` are used, making the props optional in the interface is unnecessary. Vue will automatically assign the default values to the props if they are not provided by the parent component when you use `defineProps` with default values, 

By making the props required in the interface, it explicitly stated that these props must be provided when using the component, but there are also default values.

So, it's better to keep the interface as required to ensure that the component consumers explicitly provide all the necessary props.

Reply:
> By making the props required in the interface, it explicitly stated that these props must be provided when using the component, but there are also default values.

Sorry, I still think this is contradicting :sweat_smile: If we make the props required the consumer is forced to provide them, then there is no point in defining default values... I guess the original implementation of the component had reasonable default values for a reason and you only need to provide what you want to change.

You will need to try harder to convince me in this one :stuck_out_tongue_winking_eye:

Reply:
For reference https://vuejs.org/api/sfc-script-setup.html#defineprops-defineemits says:

> the withDefaults helper provides type checks for the default values, and ensures the returned props type has the optional flags removed for properties that do have default values declared.

which seems weird at first sight, but in line with what @itisAliRH said. I think the way to think about this is that these are required; if they are not provided they are set to the default value. In a sense this does not describe what props the parent component has to pass, but what the component can expect to consume. Does that make sense ?

Reply:
(which does leave out the question whether or not you should still use `?` and that's ambiguous in the docs ... if the helper removes the optional flag maybe you should still set it ?)

Reply:
> and ensures the returned props type has the optional flags removed for properties that do have default values declared

What I understand from this sentence in the documentation is that when you use `withDefaults` you no longer need to check for undefined all over your component's code on those properties as Typescript will know it has a value even if declared as optional.

Also, that very example in the documentation has the declaration of all the props as optional and not "required with defaults" (which is what I think doesn't make sense).
```ts
export interface Props {
  msg?: string
  labels?: string[]
}

const props = withDefaults(defineProps<Props>(), {
  msg: 'hello',
  labels: () => ['one', 'two']
})
```
Does this sound more logical?

Reply:
Yes, and I think that means what you have here is correct.